plugins/cxaudio/fu-cxaudio-device.c

```diff
@@ -0,0 +1,881 @@
+fu_cxaudio_device_operation (FuCxaudioDevice *self,
+			     FuCxaudioOperation operation,
+			     FuCxaudioMemKind mem_kind,
+			     guint32 addr,
+			     guint8 *buf,
+			     guint32 bufsz,
+			     FuCxaudioOperationFlags flags,
+			     GError **error)
+{
+	const guint32 idx_read = 0x1;
+	const guint32 idx_write = 0x5;
+	const guint32 payload_max = 0x20;
+	guint32 size = 0x02800;
+	g_autoptr(GPtrArray) chunks = NULL;
+
+	g_return_val_if_fail (bufsz > 0, FALSE);
+	g_return_val_if_fail (buf != NULL, FALSE);
+
+	/* check if memory operation is supported by device */
+	if (operation == FU_CXAUDIO_OPERATION_WRITE &&
+	    mem_kind == FU_CXAUDIO_MEM_KIND_CPX_ROM) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "trying to write unwritable section %u",
+			     mem_kind);
+		return FALSE;
+	}
+
+	/* check memory address - should be within valid range */
+	switch (mem_kind) {
+	case FU_CXAUDIO_MEM_KIND_EEPROM:
+		size = 0x20000;
+		break;
+	default:
+		break;
+	}
+	if (addr > size) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "address out of range 0x%x < 0x%x",
+			     addr, size);
+		return FALSE;
+	}
+
+	/* send to hardware */
+	chunks = fu_chunk_array_new (buf, bufsz, addr, 0x0, payload_max);
+	for (guint i = 0; i < chunks->len; i++) {
+		FuChunk *chunk = g_ptr_array_index (chunks, i);
+		guint8 inbuf[FU_CXAUDIO_INPUT_REPORT_SIZE] = { 0 };
+		guint8 outbuf[FU_CXAUDIO_OUTPUT_REPORT_SIZE] = { 0 };
+
+		/* first byte is always report ID */
+		outbuf[0] = FU_CXAUDIO_MEM_WRITEID;
+
+		/* set memory address and payload length (if relevant) */
+		if (chunk->address >= 64 * 1024)
+			outbuf[1] |= 1 << 4;
+		outbuf[2] = chunk->data_sz;
+		fu_common_write_uint16 (outbuf + 3, chunk->address, G_BIG_ENDIAN);
+
+		/* set the write bit */
+		if (operation == FU_CXAUDIO_OPERATION_WRITE)
+			outbuf[1] |= 1 << 6;
+
+		/* set memtype */
+		if (mem_kind == FU_CXAUDIO_MEM_KIND_EEPROM)
+			outbuf[1] |= 1 << 5;
+
+		/* fill the report payload part */
+		if (operation == FU_CXAUDIO_OPERATION_WRITE) {
+			if (!fu_memcpy_safe (outbuf, sizeof(outbuf), idx_write, /* dst */
+					     chunk->data, chunk->data_sz, 0x0, /* src */
+					     chunk->data_sz, error))
+				return FALSE;
+		}
+		if (!fu_cxaudio_device_output_report (self, outbuf, sizeof(outbuf), error))
+			return FALSE;
+
+		/* issue additional write directive to read */
+		if (operation == FU_CXAUDIO_OPERATION_WRITE &&
+		    flags & FU_CXAUDIO_OPERATION_FLAG_VERIFY) {
+			outbuf[1] &= ~(1 << 6);
+			if (!fu_cxaudio_device_output_report (self, outbuf, sizeof(outbuf), error))
+				return FALSE;
+		}
+		if (operation == FU_CXAUDIO_OPERATION_READ ||
+		    flags & FU_CXAUDIO_OPERATION_FLAG_VERIFY) {
+			if (!fu_cxaudio_device_input_report (self,
+							     FU_CXAUDIO_MEM_READID,
+							     inbuf, sizeof(inbuf),
+							     error))
+				return FALSE;
+		}
+		if (operation == FU_CXAUDIO_OPERATION_WRITE &&
```

Comment:
I think you can also optimize here.  Right now it's:
* check if write & verify; do stuff
* check if read & verify; do stuff
* check if write & verify; do more stuff

Move the read check to the first one, and combine the two write checks

Reply:
Nah, the middle one is check if `read || verify`